cmd/registry/cmd/upload/artifact.go

```diff
@@ -101,16 +102,21 @@ func buildArtifact(ctx context.Context, parent string, filename string) (*rpc.Ar
 	return artifact, nil
 }
 
-func buildManifestArtifact(ctx context.Context, jsonBytes []byte) (*rpc.Artifact, error) {
+func buildManifestArtifact(ctx context.Context, parent string, jsonBytes []byte) (*rpc.Artifact, error) {
 	m := &rpc.Manifest{}
 	err := protojson.Unmarshal(jsonBytes, m)
 	if err != nil {
 		return nil, err
 	}
-	err = validateManifest(ctx, m)
-	if err != nil {
-		return nil, err
+	errs := controller.ValidateManifest(ctx, parent, m)
+	if len(errs) > 0 {
+		errStrings := make([]string, 0)
+		for _, err = range errs {
+			errStrings = append(errStrings, err.Error())
+		}
+		return nil, fmt.Errorf("manifest definition contains errors: %s", strings.Join(errStrings, "\n"))
```

Comment:
How does this error appear to users? I don't often see multi-line error messages or log entries.

Consider logging the errors here and return an error summarizing the issue for the main command to include in the fatal log. This might present better to users even though it pushes logging deeper into the command implementation.

```
if count := len(errs); count > 0 {
	for _, err := range errs {
		log.FromContext(ctx).WithError(err).Error("Manifest error")
	}
	return fmt.Errorf("manifest definition contains %d error(s): see logs for details", count)
}
```

Reply:
Thanks, I think this is a better approach.